returnn/tf/layers/basic.py

```diff
@@ -4769,39 +4769,53 @@ def __init__(self, eval, **kwargs):
 
 class CompareLayer(LayerBase):
   """
-  This layer performs an elementwise comparison of all provided sources, and returnns a boolean vector
-  of the same shape.
-  If the `value` parameter is provided, all elements in the provided sources are also compared to this value.
+  Compares element-wise the tokens of all input sequences among themselves and/or with a specified given value.
+  The comparisons are performed in a chain according to the order in which they are listed. 
 
-  If more than two sources are provided, each additional source will be compared against the first source,
-  and the results will be combined by a logical conjunction.
+  Example:
+    {"class": "compare", "from": ["i1", "i2"], "value": val, "kind": "less"}
+    computes i1 < i2 < val and it is true only if the whole chain of operations is true.
+    The final result is the logical "and" of all comparisons. Note that `value` is the last element to be compared to.
+  
+  A common example usage is the `end` layer in a rec subnetwork to specify the stopping criterion,
+  e.g. the last generated token is equal to the end-of-sentence token:
+
+    "output": {"class": "rec", "from": [], "unit": {
+        .
+        .
+        .
+        "end": {"class": "compare", "from": "output", "value": end_of_sentence_id}
+    }, "target": "classes0"},
   """
   layer_class = "compare"
 
   def __init__(self, kind="equal", value=None, **kwargs):
     """
-    :param str kind: Which comparison operation to use, e.g. "equal" or other supported tf comparison ops
+    :param str kind: Which comparison operation to use, e.g. "equal", "greater", "less"
+    or other supported tf comparison ops
     :param float|int|None value: if specified, will also compare to this
     """
     super(CompareLayer, self).__init__(**kwargs)
     assert len(self.sources) >= 1
     if value is None:
-      assert len(self.sources) >= 2
+      assert len(self.sources) >= 2, "{} requires at least two elements to compare.".format(self.__name__)
     op = getattr(tf, kind)  # e.g. tf.equal
     from returnn.tf.util.basic import swapaxes
-    x = self.sources[0].output.placeholder
+    common_data = Data.get_common_data([s.output for s in self.sources])
+    x = self.sources[0].output.copy_compatible_to(common_data).placeholder
     batch_axis = self.sources[0].output.batch_dim_axis
     r_last = None
-    if value is not None:
-      r_last = op(x, value)
     for source in self.sources[1:]:
-      x2 = source.output.placeholder
+      x2 = source.output.copy_compatible_to(common_data).placeholder
       if source.output.batch_dim_axis != batch_axis:
         x2 = swapaxes(x2, batch_axis, source.output.batch_dim_axis)
       r = op(x, x2)
       if r_last is not None:
         r = tf.logical_and(r_last, r)
       r_last = r
+      x = x2
+    if value is not None:
+      r_last = tf.logical_and(r_last, op(x, value))
```

Comment:
`r_last` can still be `None`. Is `tf.logical_and` capable of handling that?

Reply:
Fixed this with my last commit

Reply:
> 
> 
> The `out_type` variable is still based on the first input:
> 
> `out_type = sources[0].output.get_kwargs()`
> 
> I think this needs to be changed as well, you can have a look at the code in `CombineLayer`.

By looking at the code for `get_out_data_from_opts` from `CombineLayer` I noticed some differences that I don't know how to solve.

1. In case of error, `CombineLayer` raises a `CannotHandleUndefinedSourcesException` while `CompareLayer` returns a `Data.create_undefined`. 
2. `CombineLayer` checks if `out_type` is a dict while `CompareLayer` assumes that it is.
3. actully the point discussed, one uses the kwargs from all sources while the other only from one element. Is it related to the fact that the output is a boolean or we need to fix this?

@albertz please, give us your opinion

Reply:
`CannotHandleUndefinedSourcesException` vs `Data.create_undefined`, doesn't really matter.
`CompareLayer` does not assume that `out_type` is a dict (documentation of the type is slightly wrong for `out_type`).
The code of `CompareLayer` (including `get_out_data_from_opts`) probably is outdated, and incomplete, as you see from the whole discussion here. It never was used for these more complex cases so far, so it didn't matter.